File: oauth2/typescript/src/flows/client-credentials.ts
/**
 * Client Credentials Flow
 *
 * RFC 6749 Section 4.4 - Client Credentials Grant.
 */

import {
  ClientCredentialsParams,
  TokenResponse,
  OAuth2Config,
} from "../types";
import { OAuth2Error, ProviderError, ConfigurationError } from "../error";
import { createErrorFromResponse } from "../error/mapping";
import { HttpTransport } from "../core/transport";

/**
 * Client Credentials Flow interface.
 */
export interface ClientCredentialsFlow {
  /**
   * Request access token using client credentials.
   */
  requestToken(params?: ClientCredentialsParams): Promise<TokenResponse>;
}

/**
 * Caching Client Credentials Flow interface.
 */
export interface CachingClientCredentialsFlow extends ClientCredentialsFlow {
  /**
   * Request access token, using cache if valid.
   */
  requestToken(params?: ClientCredentialsParams): Promise<TokenResponse>;

  /**
   * Force refresh, bypassing cache.
   */
  forceRefresh(params?: ClientCredentialsParams): Promise<TokenResponse>;

  /**
   * Clear cached token.
   */
  clearCache(): void;
}

/**
 * Client Credentials Flow implementation.
 */
export class ClientCredentialsFlowImpl implements ClientCredentialsFlow {
  private config: OAuth2Config;
  private transport: HttpTransport;

  constructor(config: OAuth2Config, transport: HttpTransport) {
    this.config = config;
    this.transport = transport;

    // Client credentials flow requires client secret
    if (!config.credentials.clientSecret) {
      throw new ConfigurationError(
        "Client credentials flow requires client_secret",
        "MissingRequired"
      );
    }
  }

  async requestToken(params?: ClientCredentialsParams): Promise<TokenResponse> {
    const body = new URLSearchParams();
    body.set("grant_type", "client_credentials");

    // Scopes
    const scopes = params?.scopes ?? this.config.defaultScopes;
    if (scopes && scopes.length > 0) {
      body.set("scope", scopes.join(" "));
    }

    // Resource indicator (RFC 8707)
    if (params?.resource) {
      body.set("resource", params.resource);
    }

    // Audience
    if (params?.audience) {
      body.set("audience", params.audience);
    }

    const headers: Record<string, string> = {
      "content-type": "application/x-www-form-urlencoded",
      accept: "application/json",
    };

    // Authentication
    if (this.config.credentials.authMethod === "client_secret_basic") {
      const credentials = Buffer.from(
        `${this.config.credentials.clientId}:${this.config.credentials.clientSecret}`
      ).toString("base64");
      headers["authorization"] = `Basic ${credentials}`;
    } else {
      // Default to client_secret_post
      body.set("client_id", this.config.credentials.clientId);
      body.set("client_secret", this.config.credentials.clientSecret!);
    }

    const response = await this.transport.send({
      method: "POST",
      url: this.config.provider.tokenEndpoint,
      headers,
      body: body.toString(),
      timeout: this.config.timeout,
    });

    if (response.status !== 200) {
      throw createErrorFromResponse(response.status, response.body);
    }

    return this.parseTokenResponse(response.body);
  }

  private parseTokenResponse(body: string): TokenResponse {
    try {
      const data = JSON.parse(body);

      if (!data.access_token) {
        throw new ProviderError("Missing access_token in response", "InvalidRequest");
      }

      return {
        accessToken: data.access_token,
        tokenType: data.token_type ?? "Bearer",
        expiresIn: data.expires_in,
        scope: data.scope,
        extra: this.extractExtraFields(data),
      };
    } catch (error) {
      if (error instanceof OAuth2Error) {
        throw error;
      }
      throw new ProviderError("Invalid token response", "InvalidRequest");
    }
  }

  private extractExtraFields(data: Record<string, unknown>): Record<string, unknown> {
    const standardFields = ["access_token", "token_type", "expires_in", "scope"];
    const extra: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(data)) {
      if (!standardFields.includes(key)) {
        extra[key] = value;
      }
    }
    return Object.keys(extra).length > 0 ? extra : {};
  }
}

/**
 * Caching Client Credentials Flow implementation.
 */
export class CachingClientCredentialsFlowImpl implements CachingClientCredentialsFlow {
  private innerFlow: ClientCredentialsFlow;
  private cachedToken?: TokenResponse;
  private cachedAt?: Date;
  private refreshThreshold: number;

  constructor(
    innerFlow: ClientCredentialsFlow,
    options?: { refreshThresholdSeconds?: number }
  ) {
    this.innerFlow = innerFlow;
    this.refreshThreshold = (options?.refreshThresholdSeconds ?? 300) * 1000; // 5 minutes
  }

  async requestToken(params?: ClientCredentialsParams): Promise<TokenResponse> {
    // Check cache
    if (this.cachedToken && this.cachedAt && !this.isExpiringSoon()) {
      return this.cachedToken;
    }

    // Request new token
    const token = await this.innerFlow.requestToken(params);
    this.cachedToken = token;
    this.cachedAt = new Date();
    return token;
  }

  async forceRefresh(params?: ClientCredentialsParams): Promise<TokenResponse> {
    this.clearCache();
    return this.requestToken(params);
  }

  clearCache(): void {
    this.cachedToken = undefined;
    this.cachedAt = undefined;
  }

  private isExpiringSoon(): boolean {
    if (!this.cachedToken?.expiresIn || !this.cachedAt) {
      return true;
    }

    const expiresAt = new Date(
      this.cachedAt.getTime() + this.cachedToken.expiresIn * 1000
    );
    const threshold = new Date(Date.now() + this.refreshThreshold);
    return expiresAt <= threshold;
  }
}

/**
 * Mock Client Credentials Flow for testing.
 */
export class MockClientCredentialsFlow implements ClientCredentialsFlow {
  private requestHistory: ClientCredentialsParams[] = [];
  private nextTokenResponse?: TokenResponse;
  private nextError?: OAuth2Error;

  /**
   * Set the next token response to return.
   */
  setNextTokenResponse(response: TokenResponse): this {
    this.nextTokenResponse = response;
    return this;
  }

  /**
   * Set the next error to throw.
   */
  setNextError(error: OAuth2Error): this {
    this.nextError = error;
    return this;
  }

  /**
   * Get request history.
   */
  getRequestHistory(): ClientCredentialsParams[] {
    return [...this.requestHistory];
  }

  async requestToken(params?: ClientCredentialsParams): Promise<TokenResponse> {
    this.requestHistory.push(params ?? {});

    if (this.nextError) {
      const error = this.nextError;
      this.nextError = undefined;
      throw error;
    }

    if (this.nextTokenResponse) {
      const response = this.nextTokenResponse;
      this.nextTokenResponse = undefined;
      return response;
    }

    return {
      accessToken: "mock-client-credentials-token",
      tokenType: "Bearer",
      expiresIn: 3600,
      scope: params?.scopes?.join(" ") ?? "api",
    };
  }
}

/**
 * Create mock Client Credentials Flow for testing.
 */
export function createMockClientCredentialsFlow(): MockClientCredentialsFlow {
  return new MockClientCredentialsFlow();
}

File: oauth2/rust/src/flows/client_credentials.rs
//! Client Credentials Flow
//!
//! RFC 6749 Section 4.4 - Client Credentials Grant.

use async_trait::async_trait;
use base64::Engine;
use std::collections::HashMap;
use std::sync::Arc;

use crate::core::{HttpMethod, HttpRequest, HttpTransport};
use crate::error::{OAuth2Error, ProtocolError};
use crate::types::{ClientAuthMethod, OAuth2Config, TokenResponse};
use crate::error::create_error_from_response;

/// Client Credentials Flow request parameters.
#[derive(Debug, Clone, Default)]
pub struct ClientCredentialsRequest {
    /// Requested scopes.
    pub scopes: Option<Vec<String>>,
    /// Additional parameters.
    pub extra_params: HashMap<String, String>,
}

/// Client Credentials Flow interface.
#[async_trait]
pub trait ClientCredentialsFlow: Send + Sync {
    /// Request access token using client credentials.
    async fn request_token(
        &self,
        request: ClientCredentialsRequest,
    ) -> Result<TokenResponse, OAuth2Error>;
}

/// Client Credentials Flow implementation.
pub struct ClientCredentialsFlowImpl<T: HttpTransport> {
    config: OAuth2Config,
    transport: Arc<T>,
}

impl<T: HttpTransport> ClientCredentialsFlowImpl<T> {
    /// Create new Client Credentials Flow.
    pub fn new(config: OAuth2Config, transport: Arc<T>) -> Self {
        Self { config, transport }
    }

    fn build_request_body(&self, request: &ClientCredentialsRequest) -> String {
        let mut params = vec![("grant_type", "client_credentials".to_string())];

        // Scopes
        let scopes = request
            .scopes
            .as_ref()
            .or(Some(&self.config.default_scopes));
        if let Some(s) = scopes {
            if !s.is_empty() {
                params.push(("scope", s.join(" ")));
            }
        }

        // Client credentials in body if using post method
        if self.config.credentials.auth_method == ClientAuthMethod::ClientSecretPost {
            params.push(("client_id", self.config.credentials.client_id.clone()));
            if let Some(secret) = &self.config.credentials.client_secret {
                use secrecy::ExposeSecret;
                params.push(("client_secret", secret.expose_secret().to_string()));
            }
        }

        // Extra parameters
        for (key, value) in &request.extra_params {
            params.push((key.as_str(), value.clone()));
        }

        params
            .into_iter()
            .map(|(k, v)| format!("{}={}", urlencoding::encode(k), urlencoding::encode(&v)))
            .collect::<Vec<_>>()
            .join("&")
    }

    fn build_request_headers(&self) -> HashMap<String, String> {
        let mut headers = HashMap::new();
        headers.insert(
            "content-type".to_string(),
            "application/x-www-form-urlencoded".to_string(),
        );
        headers.insert("accept".to_string(), "application/json".to_string());

        // Add Basic auth header if using that method
        if self.config.credentials.auth_method == ClientAuthMethod::ClientSecretBasic {
            if let Some(secret) = &self.config.credentials.client_secret {
                use secrecy::ExposeSecret;
                let credentials = format!(
                    "{}:{}",
                    self.config.credentials.client_id,
                    secret.expose_secret()
                );
                let encoded = base64::engine::general_purpose::STANDARD.encode(credentials);
                headers.insert("authorization".to_string(), format!("Basic {}", encoded));
            }
        }

        headers
    }
}

#[async_trait]
impl<T: HttpTransport> ClientCredentialsFlow for ClientCredentialsFlowImpl<T> {
    async fn request_token(
        &self,
        request: ClientCredentialsRequest,
    ) -> Result<TokenResponse, OAuth2Error> {
        let body = self.build_request_body(&request);
        let headers = self.build_request_headers();

        let http_request = HttpRequest {
            method: HttpMethod::Post,
            url: self.config.provider.token_endpoint.clone(),
            headers,
            body: Some(body),
            timeout: Some(self.config.timeout),
        };

        let response = self.transport.send(http_request).await?;

        if response.status != 200 {
            return Err(create_error_from_response(response.status, &response.body));
        }

        let token_response: TokenResponse = serde_json::from_str(&response.body).map_err(|e| {
            OAuth2Error::Protocol(ProtocolError::InvalidJson {
                message: e.to_string(),
            })
        })?;

        Ok(token_response)
    }
}

/// Mock Client Credentials Flow for testing.
#[derive(Default)]
pub struct MockClientCredentialsFlow {
    request_history: std::sync::Mutex<Vec<ClientCredentialsRequest>>,
    next_token_response: std::sync::Mutex<Option<TokenResponse>>,
    next_error: std::sync::Mutex<Option<OAuth2Error>>,
}

impl MockClientCredentialsFlow {
    /// Create new mock flow.
    pub fn new() -> Self {
        Self::default()
    }

    /// Set next token response.
    pub fn set_next_token_response(&self, response: TokenResponse) -> &Self {
        *self.next_token_response.lock().unwrap() = Some(response);
        self
    }

    /// Set next error.
    pub fn set_next_error(&self, error: OAuth2Error) -> &Self {
        *self.next_error.lock().unwrap() = Some(error);
        self
    }

    /// Get request history.
    pub fn get_request_history(&self) -> Vec<ClientCredentialsRequest> {
        self.request_history.lock().unwrap().clone()
    }
}

#[async_trait]
impl ClientCredentialsFlow for MockClientCredentialsFlow {
    async fn request_token(
        &self,
        request: ClientCredentialsRequest,
    ) -> Result<TokenResponse, OAuth2Error> {
        self.request_history.lock().unwrap().push(request);

        if let Some(error) = self.next_error.lock().unwrap().take() {
            return Err(error);
        }

        if let Some(response) = self.next_token_response.lock().unwrap().take() {
            return Ok(response);
        }

        Ok(TokenResponse {
            access_token: "mock-client-credentials-token".to_string(),
            token_type: "Bearer".to_string(),
            expires_in: Some(3600),
            refresh_token: None, // Client credentials typically don't get refresh tokens
            scope: Some("api:read api:write".to_string()),
            id_token: None,
            extra: HashMap::new(),
        })
    }
}

/// Create mock Client Credentials Flow for testing.
pub fn create_mock_client_credentials_flow() -> MockClientCredentialsFlow {
    MockClientCredentialsFlow::new()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_mock_client_credentials_flow() {
        let flow = MockClientCredentialsFlow::new();

        let request = ClientCredentialsRequest {
            scopes: Some(vec!["api:read".to_string()]),
            ..Default::default()
        };

        let response = flow.request_token(request).await.unwrap();
        assert_eq!(response.access_token, "mock-client-credentials-token");
        assert!(response.refresh_token.is_none());

        let history = flow.get_request_history();
        assert_eq!(history.len(), 1);
    }

    #[tokio::test]
    async fn test_mock_with_custom_response() {
        let flow = MockClientCredentialsFlow::new();
        flow.set_next_token_response(TokenResponse {
            access_token: "custom-token".to_string(),
            token_type: "Bearer".to_string(),
            expires_in: Some(7200),
            refresh_token: None,
            scope: None,
            id_token: None,
            extra: HashMap::new(),
        });

        let request = ClientCredentialsRequest::default();
        let response = flow.request_token(request).await.unwrap();
        assert_eq!(response.access_token, "custom-token");
        assert_eq!(response.expires_in, Some(7200));
    }

    #[tokio::test]
    async fn test_mock_with_error() {
        let flow = MockClientCredentialsFlow::new();
        flow.set_next_error(OAuth2Error::Token(crate::error::TokenError::InvalidGrant {
            message: "Test error".to_string(),
        }));

        let request = ClientCredentialsRequest::default();
        let result = flow.request_token(request).await;
        assert!(result.is_err());
    }
}
